src/devices/sound/ymfm.h

```diff
@@ -0,0 +1,800 @@
+
+// ======================> ymfm_operator
+
+template<class RegisterType>
+class ymfm_operator
+{
+	enum envelope_state : u8
+	{
+		ENV_ATTACK = 0,
+		ENV_DECAY = 1,
+		ENV_SUSTAIN = 2,
+		ENV_RELEASE = 3
+	};
+
+public:
+	// constructor
+	ymfm_operator(RegisterType regs);
+
+	// register for save states
+	void save(device_t &device, u8 index);
+
+	// reset the operator state
+	void reset();
+
+	// master clocking function
+	void clock(u32 env_counter, s8 lfo_raw_pm, u16 block_freq);
+
+	// compute operator volume
+	s16 compute_volume(u16 modulation, u16 am_offset) const;
+
+	// compute volume for the OPM noise channel
+	s16 compute_noise_volume(u8 noise_state, u16 am_offset) const;
+
+	// key state control
+	void keyonoff(u8 on) { m_keyon = on; }
+	void keyon_csm() { m_csm_triggered = 1; }
+
+private:
+	// convert the generic block_freq into a 5-bit keycode
+	u8 block_freq_to_keycode(u16 block_freq);
+
+	// return the effective 6-bit ADSR rate after adjustments
+	u8 effective_rate(u8 rawrate, u8 keycode);
+
+	// start the attack phase
+	void start_attack(u8 keycode);
+
+	// start the release phase
+	void start_release();
+
+	// clock phases
+	void clock_keystate(u8 keystate, u8 keycode);
+	void clock_ssg_eg_state(u8 keycode);
+	void clock_envelope(u16 env_counter, u8 keycode);
+	void clock_phase(s8 lfo_raw_pm, u16 block_freq);
+
+	// return effective attenuation of the envelope
+	u16 envelope_attenuation(u8 am_offset) const;
+
+	// internal state
+	u32 m_phase;                     // current phase value (10.10 format)
+	u16 m_env_attenuation;           // computed envelope attenuation (4.6 format)
+	envelope_state m_env_state;      // current envelope state
+	u8 m_ssg_inverted;               // non-zero if the output should be inverted (bit 0)
+	u8 m_key_state;                  // current key state: on or off (bit 0)
+	u8 m_keyon;                      // live key on state (bit 0)
+	u8 m_csm_triggered;              // true if a CSM key on has been triggered (bit 0)
+	RegisterType m_regs;             // operator-specific registers
+};
+
+
+// ======================> ymfm_channel
+
+template<class RegisterType>
+class ymfm_channel
+{
+public:
+	// constructor
+	ymfm_channel(RegisterType regs);
+
+	// register for save states
+	void save(device_t &device, u8 index);
+
+	// reset the channel state
+	void reset();
+
+	// signal key on/off to our operators
+	void keyonoff(u8 states);
+
+	// signal CSM key on to our operators
+	void keyon_csm();
+
+	// master clocking function
+	void clock(u32 env_counter, s8 lfo_raw_pm, bool is_multi_freq);
+
+	// compute the channel output and add to the left/right output sums
+	void output(u8 lfo_raw_am, u8 noise_state, s32 &lsum, s32 &rsum, u8 rshift, s16 clipmax) const;
+
+private:
+	// convert a 6/8-bit raw AM value into an amplitude offset based on sensitivity
+	u16 lfo_am_offset(u8 am_value) const;
+
+	// internal state
+	mutable s16 m_feedback[3];            // feedback memory for operator 1
```

Comment:
Is it really a good idea to make something like this `mutable`?  If it isn’t just a cache for something, it suggests repeatedly calling a `const` member function may modify state and produce different results.

Reply:
Yeah, I'm on the fence about this. I keep the clocking step (which is non-const) separate from the output calculation step (which is const). The feedback input value for the next clock (m_feedback[2]) is an intermediate computed value from the output step. If the output is called multiple times between clocks it will produce the same m_feedback[2] value, and also doesn't consume it (it only consumes m_feedback[0] and m_feedback[1]). I could split out m_feedback[2] into a separate mutable value to make this clearer, or I could de-const the whole output chain. I'll start by doing the former, but let me know if you have a preference for the latter.

Reply:
Assuming you're ok with the current state, where I split out m_feedback_in into a separate variable.